### PR TITLE
Extended tag support

### PR DESCRIPTION
**Pull request by shokuie**
Adding extended tag support

### CODE REVIEW

File: index.js [hunk 1]
```diff
@@ -6,6 +6,22 @@ function decode(buffer) {
...
+  const form = tag >> 5 & 1;
+  let tagCode = tag & 0b11111;
+
+  if (tagCode === 0b11111) {
+    let longTag = '';
+
+    do {
+      tagCode = buffer.readUInt8(bytesRead);
+      bytesRead += 1;
+      longTag += (tagCode & 0x7f).toString(16);
```
Comment (atesgoral) [inline index.js:19]: Curious about the reason for stripping bit 8.

File: index.js [hunk 2]
```diff
@@ -6,6 +6,22 @@ function decode(buffer) {
...
+    } while (tagCode & 0b10000000);
```
Comment (atesgoral) [inline index.js:20]: Nitpick: It would be clearer to use `& 0x80` here for symmetry with the `& 0x7f` above.

File: index.js [hunk 3]
```diff
@@ -6,6 +6,22 @@ function decode(buffer) {
   let tag = buffer.readUInt8(bytesRead);
   bytesRead += 1;
 
+  const cls = tag >> 6;
+  const form = tag >> 5 & 1;
+  let tagCode = tag & 0b11111;
+
+  if (tagCode === 0b11111) {
+    let longTag = '';
```
Comment (atesgoral) [inline index.js:14]: Instead of hex string concatenation and then parsing the hex string, you can accumulate the value with bit shifts:

```
    tagCode = 0;

    do {
      byte = buffer.readUInt8(bytesRead);
      bytesRead += 1;
      tagCode = (tagCode << 8) | (byte & 0x7f);
    } while (byte & 0x80);
```

File: index.js [hunk 4]
```diff
@@ -48,7 +56,7 @@ function decode(buffer) {
   } else {
     const elements = [];
 
-    while (!elementLength || bytesRead - 2 < elementLength) {
+    while (elementLength == undefined || bytesRead - 2 < elementLength) {
```
Comment (atesgoral) [inline index.js:59]: What's the reason for this change?

File: test/index.js [hunk 5]
```diff
@@ -38,10 +38,21 @@ test('decode: tag 0', (t) => {
   );
 });
 
-test('decode: tagCode 0b11111', (t) => {
-  t.throws(() => {
-    asn1Tree.decode(b(tag(CLS_UNIVERSAL, FORM_PRIMITIVE, 0b11111), 0));
-  }, 'Extended tags are not supported');
+test('decode: tagCode 0x9F32 (high tag)', (t) => {
```
Comment (atesgoral) [inline test/index.js:41]: high tag -> extended tag?

File: index.js (hunk 3 above)
Comment (shokuie) [reply index.js]: Makes much sense, i don't know really where was I or what i was thinking when writing the code.

File: index.js (hunk 4 above)
Comment (shokuie) [reply index.js]: As far as I can remember it's to separate zero-length elements from indefinite length elements.

File: test/index.js (hunk 5 above)
Comment (shokuie) [reply test/index.js]: In ASN.1 it's called low-tag and high-tag. I believe a high-tag or extended tag is interchangeable.

File: index.js [hunk 6]
```diff
@@ -6,6 +6,20 @@ function decode(buffer) {
...
+  let tagCode = tag & 0b11111;
+
+  if (tagCode === 0b11111) {
+    tagCode = 0;
+
+    do {
+      let byte = buffer.readUInt8(bytesRead);
+      bytesRead += 1;
+      tagCode = (tagCode << 8) | (byte & 0x7f);
+    } while (tagCode & 0x80);
```
Comment (atesgoral) [inline index.js:20]: This should be `byte`

Comment (shokuie) [reply index.js:20]: I appreciate your time and dedication

File: package.json [hunk 7]
```diff
@@ -23,6 +23,9 @@
   "devDependencies": {
     "ava": "^0.22.0",
     "coveralls": "^2.13.1",
+    "eslint": "^6.8.0",
```
Comment (atesgoral) [inline package.json:26]: I assume these were added by accident

Comment (shokuie) [reply package.json:26]: Actually they are VSCode eslinter development dependency. What is your recommendation, when to add them?

Comment (atesgoral) [reply package.json:26]: I'd rather leave them out, not to be too opinioned / dependent on particular dev IDE/tooling. The other dev dependencies are for the unit tests and code coverage, and they're part of the CI.

Comment (atesgoral) [reply package.json:26]: (Don't have a very strong opinion though.)

**Comment by atesgoral** [comment]
Thanks a lot for adding this support BTW!

File: test/index.js (hunk 5 above)
Comment (atesgoral) [reply test/index.js]: Good to know!

**Comment by atesgoral** [comment]
@shokuie Published v0.1.0. Thanks for this!
